README.md

<!--[![Build Status](https://travis-ci.com/sbohora/saucpy.svg?token=shyYTzvvbsLRHsRAWXTg&branch=master)](https://travis-ci.com/sbohora/saucpy) -->


## Semi-parametric Area Under the Curve (sAUC) Regression
Perform AUC analyses with discrete covariates and a semi-parametric estimation

### Model

We consider applications that compare a response variable y between two groups (A and B) while adjusting for k categorical covariates ![](http://latex.codecogs.com/gif.latex?X_1,X_2,...,X_k).  The response variable y is a continuous or ordinal variable that is not normally distributed.  Without loss of generality, we assume each covariate is coded such that ![](http://latex.codecogs.com/gif.latex?X_i%3D1,...,n_i),for ![](http://latex.codecogs.com/gif.latex?i%3D1,...,k). For each combination of the levels of the covariates, we define the Area Under the ROC curve (AUC) in the following way:

![](http://latex.codecogs.com/gif.latex?%5Cpi_%7Bx_1%20x_2...x_k%7D%3DP(Y%5EA%3EY%5EB%7CX_1%3Dx_1,X_2%3Dx_2,...,X_k%3Dx_k%20)+%5Cfrac%7B1%7D%7B2%7D%20P(Y%5EA%3DY%5EB%7CX_1%3Dx_1,X_2%3Dx_2,...,X_k%3Dx_k%20),)

where ![](http://latex.codecogs.com/gif.latex?x_1%3D1,...,n_1,...,x_k%3D1,...,n_k), and ![](http://latex.codecogs.com/gif.latex?Y%5EA) and ![](http://latex.codecogs.com/gif.latex?Y%5EB) are two randomly chosen observations from Group A and B, respectively.  The second term in the above equation is for the purpose of accounting ties.

For each covariate ![](http://latex.codecogs.com/gif.latex?X_i), without loss of generality, we use the last category as the reference category and define (![](http://latex.codecogs.com/gif.latex?n_i-1)) dummy variables ![](http://latex.codecogs.com/gif.latex?X_i%5E%7B(1)%7D,X_i%5E%7B(2)%7D,...,X_i%5E%7B(n_i-1)%7D) such that 

![](http://latex.codecogs.com/gif.latex?X_i%5E%7B(j)%7D%20(x)%3D%201,%20if%20j%20%3D%20x) and ![](http://latex.codecogs.com/gif.latex?0,%20if%20j%20%5Cne%20x.)

where ![](http://latex.codecogs.com/gif.latex?i%3D1,...,k;%20j%3D1,...,n_i-1;%20x%3D1,...,n_i).   We model the association between AUC ![](http://latex.codecogs.com/gif.latex?%5Cpi_(x_1%20x_2...x_k)) and covariates using a logistic model.  Such a model specifies that the logit of ![](http://latex.codecogs.com/gif.latex?%5Cpi_(x_1%20x_2...x_k)) is a linear combination of terms that are products of the dummy variables defined above.  Specifically,

![](http://latex.codecogs.com/gif.latex?logit(%5Cpi_%7Bx_1%20x_2...x_k%20%7D%20)%3DZ_%7B(x_1%20x_2...x_k%20)%7D%20%5Cboldsymbol%7B%5Cbeta%7D,) 

where ![](http://latex.codecogs.com/gif.latex?Z_%7B(x_1%20x_2...x_k)%7D) is a row vector whose elements are zeroes or ones and are products of ![](http://latex.codecogs.com/gif.latex?X_1%5E%7B(1)%7D%20(x_1%20),...,X_1%5E%7B(n_i-1)%20%7D%20(x_1),...,X_k%5E%7B(1)%7D%20(x_k),...,X_k%5E%7B(n_k-1)%7D%20(x_k)), and ![](http://latex.codecogs.com/gif.latex?%5Cboldsymbol%7B%5Cbeta%7D) is a column vector of nonrandom unknown parameters.  Now, define a column vector ![](http://latex.codecogs.com/gif.latex?%5Cpi) by stacking up ![](http://latex.codecogs.com/gif.latex?%5Cpi_(x_1%20x_2...x_k%20)) and define a matrix Z by stacking up ![](http://latex.codecogs.com/gif.latex?Z_%7B(x_1%20x_2...x_k%20)%7D), as ![](http://latex.codecogs.com/gif.latex?x_i) ranges from 1 to ![](http://latex.codecogs.com/gif.latex?n_i,%20i%3D1,...,k), our final model is  

![](http://latex.codecogs.com/gif.latex?logit(%5Cpi)%3DZ%5Cboldsymbol%7B%5Cbeta%7D%20...(1))

The reason for us to use a logit transformation of the AUC instead of using the original AUC is for variance stabilization.  We will illustrate the above general model using examples.


### Estimation

First, we denote the number of observations with covariates ![](http://latex.codecogs.com/gif.latex?X_1%3Di_1,...,X_k%3Di_k) in groups A and B by ![](http://latex.codecogs.com/gif.latex?N_(i_1...i_k)%5EA) and ![](http://latex.codecogs.com/gif.latex?N_(i_1...i_k)%5EB), respectively.  We assume both ![](http://latex.codecogs.com/gif.latex?N_(i_1...i_k)%5EA) and ![](http://latex.codecogs.com/gif.latex?N_(i_1...i_k)%5EB) are greater than zero in the following development.  An unbiased estimator of ![](http://latex.codecogs.com/gif.latex?%5Cpi_(i_1...i_k)) proposed by Mann and Whitney (1947) is

![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Cpi%7D_(i_1...i_k)%3D%5Cfrac%7B%5Csum_%7Bl%3D1%7D%5E%7BN_%7Bi_1...i_k%7D%5EA%7D%20%5Csum_%7Bj%3D1%7D%5E%7BN_%7Bi_1...i_k%7D%5EB%7D%20I_%7Blj%7D%7D%7BN_%7Bi_1...i_k%7D%5EA%20N_%7Bi_1...i_k%7D%5EB%7D,)

where 

![](http://latex.codecogs.com/gif.latex?I_(i_1...i_k);%20lj%20%3D%201) if ![](http://latex.codecogs.com/gif.latex?Y_%7Bi_1...i_k;l%7D%5EA%20%3E%20Y_%7Bi_1...i_k;j%7D%5EB)

and

![](http://latex.codecogs.com/gif.latex?I_(i_1...i_k);%20lj%20%3D%20%5Cfrac%7B1%7D%7B2%7D) if ![](http://latex.codecogs.com/gif.latex?Y_%7Bi_1...i_k;l%7D%5EA%20%3D%20Y_%7Bi_1...i_k;j%7D%5EB)

and

![](http://latex.codecogs.com/gif.latex?I_(i_1...i_k);%20lj%20%3D%200) if ![](http://latex.codecogs.com/gif.latex?Y_%7Bi_1...i_k;l%7D%5EA%20%3C%20Y_%7Bi_1...i_k;j%7D%5EB)


and ![](http://latex.codecogs.com/gif.latex?Y_(i_1...i_k;%20l)%5EA) and ![](http://latex.codecogs.com/gif.latex?Y_(i_1...i_k;%20j)%5EB) are observations with ![](http://latex.codecogs.com/gif.latex?X_1%3Di_1,...,X_k%3Di_k) in groups A and B, respectively.  Delong, Delong and Clarke-Pearson (1988) have shown that 

![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Cpi%7D_%7Bi_1...i_k%7D%20%5Capprox%20N(%5Cpi_%7Bi_1...i_k%7D,%5Csigma_%7Bi_1...i_k%7D%5E2)).	

In order to obtain an estimator for ![](http://latex.codecogs.com/gif.latex?%5Csigma_%7Bi_1...i_k%7D%5E2), they first computed

![](http://latex.codecogs.com/gif.latex?V_%7Bi_1...i_k;%20l%7D%5EA%3D%5Cfrac%7B1%7D%7BN_%7Bi_1...i_k%7D%5EB%20%7D%20%5Csum_%7Bj%3D1%7D%5E%7BN_%7Bi_1...i_k%7D%5EB%7D%20I_%7Blj%7D,%20%20%09l%3D1,...,N_%7Bi_1...i_k%7D%5EA)

and

![](http://latex.codecogs.com/gif.latex?V_%7Bi_1...i_k;j%7D%5EB%3D%5Cfrac%7B1%7D%7BN_%7Bi_1...i_k%7D%5EA%20%7D%20%5Csum_%7Bl%3D1%7D%5E%7BN_%7Bi_1...i_k%7D%5EA%7D%20I_%7Blj%7D,%20%20%09j%3D1,...,N_%7Bi_1...i_k%7D%5EB).

Then, an estimate of the variance of the nonparametric AUC was

![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Csigma%7D_%7Bi_1...i_k%7D%5E2%3D%5Cfrac%7B(s_%7Bi_1...i_k%7D%5EA%20)%5E2%7D%7BN_%7Bi_1...i_k%7D%5EA%7D%20+%20%5Cfrac%7B(s_%7Bi_1...i_k%7D%5EB%20)%5E2%7D%7BN_%7Bi_1...i_k%7D%5EB%7D),

where 

![](http://latex.codecogs.com/gif.latex?(s_%7Bi_1...i_k%7D%5EA%20)%5E2) and ![](http://latex.codecogs.com/gif.latex?(s_%7Bi_1...i_k%7D%5EB%20)%5E2) were the sample variances of 

![](http://latex.codecogs.com/gif.latex?V_%7Bi_1...i_k;%20l%7D%5EA;%20l%3D1,...,N_%7Bi_1...i_k%7D%5EA) and ![](http://latex.codecogs.com/gif.latex?V_%7Bi_1...i_k;%20j%7D%5EB;%20j%3D1,...,N_%7Bi_1...i_k%7D%5EB,) respectively.  Clearly, we need both ![](http://latex.codecogs.com/gif.latex?N_%7Bi_1...i_k%7D%5EA) and ![](http://latex.codecogs.com/gif.latex?N_%7Bi_1...i_k%7D%5EB) are greater than two in order to compute ![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Csigma%7D_%7Bi_1...i_k%7D%5E2).

Now, in order to estimate parameters in Model (1), we first derive the asymptotic variance of ![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Cgamma%7D_%7Bi_1...i_k%7D) using the delta method, which results in

![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Cgamma%7D_%7Bi_1...i_k%7D%3Dlogit(%5Chat%7B%5Cpi%7D_%7Bi_1...i_k%7D)%20%5Capprox%20N(logit(%5Cpi_%7Bi_1...i_k%7D),%5Ctau_%7Bi_1...i_k%7D%5E2),)

where ![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Ctau%7D_%7Bi_1...i_k%7D%5E2%3D%5Cfrac%7B%5Chat%7B%5Cgamma%7D_%7Bi_1...i_k%7D%5E2%7D%7B%5Chat%7B%5Cpi%7D_%7Bi_1...i_k%7D%5E2%20%20(1-%5Chat%7B%5Cpi%7D_%7Bi_1...i_k%7D)%5E2%7D) 

Rewriting the above model, we obtain

![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Cgamma%7D_%7Bi_1...i_k%7D%3Dlogit(%5Cpi_%7Bi_1...i_k%20%7D)%20%3DZ_%7Bi_1...i_k%7D%20%5Cboldsymbol%7B%5Cbeta%7D%20+%20%5Cepsilon_%7Bi_1...i_k%7D)
         
where, 

![](http://latex.codecogs.com/gif.latex?%5Cepsilon_%7Bi_1,...,i_k%7D%20%5Capprox%20N(0,%5Ctau_%7Bi_1,...,i_k%7D%5E2)).  Then, by stacking up the ![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Cgamma%7D_%7B1_i,...,i_k%7D) to be 
![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Cgamma%7D,%20Z_%7Bi_1...i_k%7D) to be ![](http://latex.codecogs.com/gif.latex?%5Cboldsymbol%7BZ%7D), and ![](http://latex.codecogs.com/gif.latex?%5Cepsilon_%7Bi_1,...,i_k%7D) to be 
![](http://latex.codecogs.com/gif.latex?%5Cboldsymbol%7B%5Cepsilon%7D), we have

![](http://latex.codecogs.com/gif.latex?%5Cboldsymbol%7B%5Chat%7B%5Cgamma%7D%7D%20%3Dlogit%20%5Cboldsymbol%7B%5Chat%7B%5Cpi%7D%7D%20%3D%20%5Cboldsymbol%7BZ%5Cbeta%20+%20%5Cepsilon%7D), 

where, ![](http://latex.codecogs.com/gif.latex?E(%5Cepsilon)%3D0) and ![](http://latex.codecogs.com/gif.latex?%5Chat%7BT%7D%3DVar(%5Cepsilon)%3Ddiag(%5Chat%7B%5Ctau%7D_%7Bi_1...%20i_k%7D%5E2)) which is a diagonal matrix.  Finally, by using the generalized least squares method, we estimate the parameters β  and its variance-covariance matrix as follows;

![](http://latex.codecogs.com/gif.latex?%5Cboldsymbol%7B%5Chat%7B%5Cbeta%7D%20%3D%7B(%5Chat%7BZ%7D%5ET%20%20%5Chat%7BT%7D%5E%7B-1%7D%20%20Z)%7D%5E%7B-1%7D%20Z%5ET%20%20%5Chat%7BT%7D%5E%7B-1%7D%20%5Chat%7B%5Cgamma%7D%7D)		

and
![](http://latex.codecogs.com/gif.latex?%5Chat%7BV%7D(%5Cboldsymbol%7B%5Chat%7B%5Cbeta%7D%7D)%20%3D%20%5Cboldsymbol%7B%7B(%5Chat%7BZ%7D%5ET%20%20%5Chat%7BT%7D%5E%7B-1%7D%20%20Z)%7D%5E%7B-1%7D%7D)

The above equations can be used to construct a 100(1-α)% Wald confidence intervals for ![](http://latex.codecogs.com/gif.latex?%5Cboldsymbol%7B%5Cbeta_i%7D) using formula

![](http://latex.codecogs.com/gif.latex?%5Chat%7B%5Cbeta%7D_i%20%5Cpm%20Z_%7B1-%5Cfrac%7B%5Calpha%7D%7B2%7D%7D%20%5Csqrt%7B%5Chat%7BV%7D(%5Chat%7B%5Cbeta%7D_i)%7D),

where ![](http://latex.codecogs.com/gif.latex?Z_%7B1-%5Cfrac%7B%5Calpha%7D%7B2%7D%7D) is the ![](http://latex.codecogs.com/gif.latex?(1-%5Cfrac%7B%5Calpha%7D%7B2%7D)%5E%7Bth%7D) quantile of the standard normal distribution.  Equivalently, we reject 

![](http://latex.codecogs.com/gif.latex?H_0:%5Cbeta_i%20%3D%200)  if ![](http://latex.codecogs.com/gif.latex?%7C%5Chat%7B%5Cbeta%7D_i%7C%20%3E%20Z_%7B1-%5Cfrac%7B%5Calpha%7D%7B2%7D%7D%20%5Csqrt%7B%5Chat%7BV%7D(%5Chat%7B%5Cbeta%7D_i)%7D.),

The p-value for testing ![](http://latex.codecogs.com/gif.latex?H_0) is ![](http://latex.codecogs.com/gif.latex?2%20*%20P(Z%20%3E%20%7C%5Chat%7B%5Cbeta%7D_i%7C/%5Csqrt%7B%5Chat%7BV%7D(%5Chat%7B%5Cbeta%7D_i)%7D),)


, where Z is a random variable with the standard normal distribution.
	Now, the total number of cells (combinations of covariates ![](http://latex.codecogs.com/gif.latex?X_1,...,X_k)) is ![](http://latex.codecogs.com/gif.latex?n_1%20n_2...n_k). As mentioned earlier, for a cell to be usable in the estimation, the cell needs to have at least two observations from Group A and two observations from Group B.  As long as the total number of usable cells is larger than the dimension of ![](http://latex.codecogs.com/gif.latex?%5Cboldsymbol%7B%5Cbeta%7D), then the matrix ![](http://latex.codecogs.com/gif.latex?%7B%5Cboldsymbol%7B%5Chat%7BZ%7D%5ET%20%20%5Chat%7BT%7D%5E%7B-1%7D%20%20Z%7D%7D) is invertible and consequently,![](http://latex.codecogs.com/gif.latex?%5Cboldsymbol%7B%5Chat%7B%5Cbeta%7D%7D) is computable and model (1) is identifiable.



HISTORY.md
## History

### 0.0.1

- [FIX] Work on new version of saucpy package


tests/test.py
from pandas import read_csv
from saucpy import sAUC
#import saucpy

# Data analysis
fasd = read_csv("../data/fasd.csv")
fasd['group'] = fasd['group'].astype('category')
fasd['x1']    = fasd['x1'].astype('category')
fasd['x2']    = fasd['x2'].astype('category')

ds = read_csv("../data/ds.csv")
ds['group'] = ds['group'].astype('category')
ds['x1']    = ds['x1'].astype('category')
ds['x2']    = ds['x2'].astype('category')
ds['x3']    = ds['x3'].astype('category')

sAUC.semiparametricAUC(response = "y", treatment_group = ["group"], input_covariates = ["x1"], data = fasd)
sAUC.semiparametricAUC(response = "y", treatment_group = ["group"], input_covariates = ["x1","x2"], data = fasd)
sAUC.semiparametricAUC(response = "y", treatment_group = ["group"], input_covariates = ["x1","x2", "x3"], data = ds)


tests/saucpy.py
from pandas import read_csv
from saucpy import sAUC
#import saucpy

def main():
    p1 = sAUC.calculate_auc([2,.3,.4,2,1.2], [0.2, 1,2,1,2,3])
    print(p1)
    
if __name__ == '__main__':
    main()

tests/example.py
from pandas import read_csv
from saucpy import sAUC

# Data analysis
fasd = read_csv("../saucpy/data/fasd.csv")
fasd['group'] = fasd['group'].astype('category')
fasd['x1']    = fasd['x1'].astype('category')
fasd['x2']    = fasd['x2'].astype('category')

ds = read_csv("../saucpy/data/ds.csv")
ds['group'] = ds['group'].astype('category')
ds['x1']    = ds['x1'].astype('category')
ds['x2']    = ds['x2'].astype('category')
ds['x3']    = ds['x3'].astype('category')

sAUC.semiparametricAUC(response = "y", treatment_group = ["group"], input_covariates = ["x1"], data = fasd)
sAUC.semiparametricAUC(response = "y", treatment_group = ["group"], input_covariates = ["x1","x2"], data = fasd)
sAUC.semiparametricAUC(response = "y", treatment_group = ["group"], input_covariates = ["x1","x2", "x3"], data = ds)


saucpy/saucpy.py
import numpy
from pandas import DataFrame
from patsy import dmatrix
from scipy.stats import norm

class sAUC(object):
    """An :class:`sAUC <saucpy.sAUC>` object."""
    """ This class contains methods to perform sAUC model."""
    def calculate_auc(ya, yb):
        """This function calculates different estimates related to AUC. 
        This function calculates variance of predicted AUC, logit of predicted AUC,
        and variance of logit of predicted AUC responses passed.
        
        :param ya: a list of numeric responses
        :param yb: a list of numeric responses
        
        Usage::
            >>> calculate_auc(ya=[2,0.4,3.6,2.41], yb= [1.2,0.4,1.6,1.5])
        """
        m = len(ya)
        p = len(yb)
        I = numpy.zeros(shape=(m, p))
        for i in range(m):
            for j in range(p):
                if ya[i] > yb[j]:
                    I[i, j] = 1
                elif ya[i] == yb[j]:
                    I[i, j] = 0.5
                else:
                    I[i, j] = 0
        #finv = lambda x: (-(numpy.log((1/x)-1)))
        auchat = numpy.mean(I)
        #finvhat = finv(auchat)
        vya = numpy.apply_along_axis(numpy.mean, 1, I)
        vyb = numpy.apply_along_axis(numpy.mean, 0, I)
        svarya = numpy.var(vya)
        svaryb = numpy.var(vyb)
        vhat_auchat = (svarya / m) + (svaryb / p)
        #v_finv_auchat = vhat_auchat/((auchat**2)*(1-auchat)**2)
        logitauchat = numpy.log(auchat / (1 - auchat))
        var_logitauchat = vhat_auchat / ((auchat**2) * (1 - auchat)**2)
        return([var_logitauchat, logitauchat])

    #print(calculate_auc(ya=[2,0.4,3.6,2.41], yb= [1.2,0.4,1.6,1.5]))
        
    # expand.grid
    def expand_grid(*itrs):
        def product(*args, **kwds):
            pools = map(tuple, args) * kwds.get('repeat', 1)
            result = [[]]
            for pool in pools:
                result = [x+[y] for x in result for y in pool]
                for prod in result:
                    yield tuple(prod)
        new_product = list(product(*itrs))
        return {'Var{}'.format(i + 1): [x[i]
                    for x in new_product]
                for i in range(len(itrs))}

#    expand_grid([1, 2, 3], [2, 1])

    def semiparametricAUC(response, treatment_group, input_covariates, data):
        assert response is not None, "Argument response is missing."
        assert treatment_group is not None, "Argument treatment_group is missing."
        assert input_covariates is not None, "Argument input_covariates is missing. Please put covariates as list. For e.g. ['x1','x2']"
        assert data is not None, "Argument data is missing. Please, specify name of pandas DataFrame."

        print("Data are being analyzed. Please, be patient.\n\n")

        d = DataFrame(data)
        group_covariates = treatment_group + input_covariates

        # split
        grouped_d = d.groupby(group_covariates)[response]

        keys = list((grouped_d.groups.keys()))

        def calculate_auc(ya, yb):
            m = len(ya)
            p = len(yb)
            I = numpy.zeros(shape=(m, p))
            for i in range(m):
                for j in range(p):
                    if ya[i] > yb[j]:
                        I[i, j] = 1
                    elif ya[i] == yb[j]:
                        I[i, j] = 0.5
                    else:
                        I[i, j] = 0
            #finv = lambda x: (-(numpy.log((1/x)-1)))
            auchat = numpy.mean(I)
            #finvhat = finv(auchat)
            vya = numpy.apply_along_axis(numpy.mean, 1, I)
            vyb = numpy.apply_along_axis(numpy.mean, 0, I)
            svarya = numpy.var(vya)
            svaryb = numpy.var(vyb)
            vhat_auchat = (svarya / m) + (svaryb / p)
            #v_finv_auchat = vhat_auchat/((auchat**2)*(1-auchat)**2)
            logitauchat = numpy.log(auchat / (1 - auchat))
            var_logitauchat = vhat_auchat / ((auchat**2) * (1 - auchat)**2)
            return([var_logitauchat, logitauchat])

        dict_df = {}
        auchat_container = {}
        my_card_1 = {}
        for i in range(len(keys)):
            # print(i)
            dict_df[i] = d.loc[grouped_d.groups[keys[i]]]

        for j in range(int(0.5 * len(dict_df))):
            # print(j)
            auchat_container[j], my_card_1[j] = (calculate_auc(dict_df[j].loc[:, response].tolist(), 
                            dict_df[j + int(0.5 * len(dict_df))].loc[:, response].tolist()))

        var_logitauchat = [v for v in auchat_container.values()]
        gamma1 = [v for v in my_card_1.values()]

        # get levels
        df_keys = DataFrame(keys)
        df_keys.columns = group_covariates
        ds_only_covariates = df_keys[input_covariates]

        select_row = int(0.5 * len(ds_only_covariates))

        ds_expand = ds_only_covariates[:select_row]

        #ds_levels = {}
        # for i in input_covariates:
        #   ds_levels[i] = (d[i].cat.categories)

        #ds_expand = (DataFrame(expand_grid(*ds_levels.values())))

        def convert_to_factor(df):
            df = DataFrame(df)
            for i in df.columns:
                df[i] = df[i].astype('category')
            return(df)

        ds_expand = convert_to_factor(ds_expand)
        #ds_expand.columns = input_covariates

        var_list = '+'.join(input_covariates)

        #from patsy import *
        # model.matrix
        Z = (dmatrix(var_list, ds_expand))

        # get levels
        di = Z.design_info

        Z.columns = di.column_names
        #Z.sort('x2', ascending = [True])

        tau = numpy.diag([1 / i for i in var_logitauchat])

        #from numpy.linalg import inv
        ztauz = numpy.linalg.inv(Z.T.dot(tau).dot(Z))

        var_betas = numpy.diag(ztauz)
        std_error = numpy.sqrt(var_betas)
        betas = ztauz.dot(Z.T).dot(tau).dot(gamma1)

        #from scipy.stats import norm
        threshold = norm.ppf(0.975)

        lo = betas - threshold * std_error
        up = betas + threshold * std_error

        p_values = (norm.cdf(-numpy.abs(betas), loc=0, scale=std_error)) * 2

        results = DataFrame(numpy.vstack((betas, std_error, lo, up, p_values)).T)
        results.columns = ["Coefficients", "Std. Error", "2.5%", "97.5%", "Pr(>|z|)"]
        results.index = di.column_names
        print("\nModel Summary")
        return(results)

sAUCpy/calculate-auc.py
# -*- coding: utf-8 -*-
'''
Provides the :class:`SAUC <sAUCpy.sAUCpy.SAUC>` class, a semi-parametric AUC method``.
'''
from numpy import *

class sAUCpy(object):
    """
    Semi-parametric Area Under the Curve (AUC) Regreesion model.

    Parameters
    ----------
    fit_intercept : boolean, optional
        whether to calculate the intercept for this model. If set
        to false, no intercept will be used in calculations
        (e.g. data is expected to be already centered).

    normalize : boolean, optional, default False
        If True, the regressors X will be normalized before regression.
        This parameter is ignored when `fit_intercept` is set to False.
        When the regressors are normalized, note that this makes the
        hyperparameters learnt more robust and almost independent of the number
        of samples. The same property is not valid for standardized data.
        However, if you wish to standardize, please use
        `preprocessing.StandardScaler` before calling `fit` on an estimator
        with `normalize=False`.

    copy_X : boolean, optional, default True
        If True, X will be copied; else, it may be overwritten.

    n_jobs : int, optional, default 1
        The number of jobs to use for the computation.
        If -1 all CPUs are used. This will only provide speedup for
        n_targets > 1 and sufficient large problems.

    Attributes
    ----------
    coef_ : array, shape (n_features, ) or (n_targets, n_features)
        Estimated coefficients for the linear regression problem.
        If multiple targets are passed during the fit (y 2D), this
        is a 2D array of shape (n_targets, n_features), while if only
        one target is passed, this is a 1D array of length n_features.

    residues_ : array, shape (n_targets,) or (1,) or empty
        Sum of residuals. Squared Euclidean 2-norm for each target passed
        during the fit. If the linear regression problem is under-determined
        (the number of linearly independent rows of the training matrix is less
        than its number of linearly independent columns), this is an empty
        array. If the target vector passed during the fit is 1-dimensional,
        this is a (1,) shape array.

        .. versionadded:: 0.18

    intercept_ : array
        Independent term in the linear model.

    Notes
    -----
    From the implementation point of view, this is just plain Ordinary
    Least Squares (scipy.linalg.lstsq) wrapped as a predictor object.

    """

    def __init__(self):
        print
        "This is init."

    # def finv(x):
    # return(-log((1/x)-1))
    #
    # def finv(x):
    # return(-log((1/x)-1))
    
    finv = lambda x: (-log((1/x)-1))
       
    m = len(ya)
    p = len(yb)
    I = numpy.zeros(shape = (m, p))
    
    def calculate_auc(ya, yb, data = None):
        for i in range(m):
            for j in range(p):
                if ya[i] > yb[j]: I[i,j] = 1
                elif ya[i] == yb[j]: I[i,j] = 0.5
                else: I[i,j] = 0
                auchat = numpy.mean(I)
                finvhat = finv(auchat)
                vya = numpy.apply_along_axis(numpy.mean, 1, I)
                vyb = numpy.apply_along_axis(numpy.mean, 0, I)
                svarya = numpy.var(vya)
                svaryb = numpy.var(vyb)
                vhat_auchat = (svarya/m) + (svaryb/p)
                v_finv_auchat = vhat_auchat/((auchat**2)*(1-auchat)**2)
                logitauchat = log(auchat/(1-auchat))
                var_logitauchat = vhat_auchat /((auchat**2)*(1-auchat)**2)
                return(var_logitauchat)
    
    ya = [3,2,4,3]
    yb = [1,2,2,3]
    
    calculate_auc(ya=ya, yb= yb)






sAUCpy/sAUC.py
import numpy
from pandas import DataFrame, read_csv
from patsy import dmatrix
from scipy.stats import norm
from itertools import product

def calculate_auc(ya, yb):
    m = len(ya)
    p = len(yb)
    I = numpy.zeros(shape = (m, p))
    for i in range(m):
        for j in range(p):
            if ya[i] > yb[j]:
                I[i,j] = 1
            elif ya[i] == yb[j]:
                I[i,j] = 0.5
            else:
                I[i,j] = 0
    #finv = lambda x: (-(numpy.log((1/x)-1)))
    auchat = numpy.mean(I)
    #finvhat = finv(auchat)
    vya = numpy.apply_along_axis(numpy.mean, 1, I)
    vyb = numpy.apply_along_axis(numpy.mean, 0, I)
    svarya = numpy.var(vya)
    svaryb = numpy.var(vyb)
    vhat_auchat = (svarya/m) + (svaryb/p)
    #v_finv_auchat = vhat_auchat/((auchat**2)*(1-auchat)**2)
    logitauchat = numpy.log(auchat/(1-auchat))
    var_logitauchat = vhat_auchat /((auchat**2)*(1-auchat)**2)
    return([var_logitauchat, logitauchat])

#print(calculate_auc(ya=[2,0.4,3.6,2.41], yb= [1.2,0.4,1.6,1.5]))

#expand.grid
def expand_grid(*itrs):
   new_product = list(product(*itrs))
   return {'Var{}'.format(i+1):[x[i]
        for x in new_product]
            for i in range(len(itrs))}

expand_grid([1,2,3],[2,1])

def sauc(response, treatment_group, input_covariates, data):
    assert response is not None, "Argument response is missing."
    assert treatment_group is not None, "Argument treatment_group is missing."
    assert input_covariates is not None, "Argument input_covariates is missing. Please put covariates as list. For e.g. ['x1','x2']"
    assert data is not None, "Argument data is missing. Please, specify name of pandas DataFrame."

    print("Data are being analyzed. Please, be patient.\n\n")

    d = DataFrame(data)
    group_covariates = treatment_group + input_covariates

    #split
    grouped_d = d.groupby(group_covariates)[response]

    keys = list((grouped_d.groups.keys()))

    dict_df = {}
    auchat_container = {}
    my_card_1 = {}
    for i in range(len(keys)):
        #print(i)
        dict_df[i] = d.loc[grouped_d.groups[keys[i]]]

    for j in range(int(0.5 * len(dict_df))):
        #print(j)
        auchat_container[j], my_card_1[j] = (calculate_auc(dict_df[j].loc[:,response].tolist(),dict_df[j + int(0.5 * len(dict_df))].loc[:,response].tolist()))

    var_logitauchat = [ v for v in auchat_container.values() ]
    gamma1 = [ v for v in my_card_1.values() ]

    # get levels
    df_keys = DataFrame(keys)
    df_keys.columns = group_covariates
    ds_only_covariates = df_keys[input_covariates]

    select_row = int(0.5*len(ds_only_covariates))

    ds_expand = ds_only_covariates[:select_row]

    #ds_levels = {}
    #for i in input_covariates:
     #   ds_levels[i] = (d[i].cat.categories)

    #ds_expand = (DataFrame(expand_grid(*ds_levels.values())))

    def convert_to_factor(df):
        df = DataFrame(df)
        for i in df.columns:
            df[i] = df[i].astype('category')
        return(df)

    ds_expand = convert_to_factor(ds_expand)
    #ds_expand.columns = input_covariates

    var_list = '+'.join(input_covariates)

    #from patsy import *
    #model.matrix
    Z = (dmatrix(var_list, ds_expand))

    #get levels
    di = Z.design_info

    Z.columns = di.column_names
    #Z.sort('x2', ascending = [True])

    tau  =  numpy.diag([1/i for i in var_logitauchat])

    #from numpy.linalg import inv
    ztauz = numpy.linalg.inv(Z.T.dot(tau).dot(Z))

    var_betas = numpy.diag(ztauz)
    std_error = numpy.sqrt(var_betas)
    betas = ztauz.dot(Z.T).dot(tau).dot(gamma1)

    #from scipy.stats import norm
    threshold = norm.ppf(0.975)

    lo = betas - threshold*std_error
    up = betas + threshold*std_error

    p_values = (norm.cdf(-numpy.abs(betas), loc = 0, scale = std_error))*2

    results = DataFrame(numpy.vstack((betas,std_error, lo,up,p_values)).T)
    results.columns = ["Coefficients","Std. Error", "2.5%", "97.5%", "Pr(>|z|)"]
    results.index = di.column_names
    print("\nModel Summary")
    return(results)

# Data analysis
fasd = read_csv("../data/fasd.csv")
fasd['group'] = fasd['group'].astype('category')
fasd['x1']    = fasd['x1'].astype('category')
fasd['x2']    = fasd['x2'].astype('category')

ds = read_csv("../data/ds.csv")
ds['group'] = ds['group'].astype('category')
ds['x1']    = ds['x1'].astype('category')
ds['x2']    = ds['x2'].astype('category')
ds['x3']    = ds['x3'].astype('category')

sauc(response = "y", treatment_group = ["group"], input_covariates = ["x1"], data = fasd)
sauc(response = "y", treatment_group = ["group"], input_covariates = ["x1","x2"], data = fasd)

sauc(response = "y", treatment_group = ["group"], input_covariates = ["x1","x2", "x3"], data = ds)
